model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/entities/AuthenticatedClientSessionEntity.java

```diff
@@ -83,6 +85,28 @@ public void setTimestamp(int timestamp) {
         this.timestamp = timestamp;
     }
 
+    public int getUserSessionStarted() {
+        String started = getNotes().get(AuthenticatedClientSessionModel.USER_SESSION_STARTED_AT_NOTE);
+        return started == null ? timestamp : Integer.parseInt(started);
```

Comment:
This defaults to `timestamp` while `AuthenticatedClientSessionModel.getUserSessionStarted` defaults to 0 if `USER_SESSION_STARTED_AT_NOTE` is not set. Is it intentional?

Reply:
Yes, it's correct. Previously `getTimestamp` was always used. So this is maintaining the same behavior just in case. If we use 0 all the session without the note would be expired. Note that we are doing the same in the map part ([here](https://github.com/keycloak/keycloak/blob/510cc404c531b9faa4f7dacc5817e6e7d38971ef/model/map/src/main/java/org/keycloak/models/map/userSession/SessionExpiration.java#L32)).

Reply:
But why we don't use `timestamp` in `AuthenticatedClientSessionModel` but 0 instead?

```
default int getUserSessionStarted() {
        String started = getNote(USER_SESSION_STARTED_AT_NOTE);
        return started == null ? 0 : Integer.parseInt(started);
}
```
and older code

```
default int getStarted() {
        String started = getNote(STARTED_AT_NOTE);
        // Fallback to 0 if "started" note is not available. This can happen for the offline sessions migrated from old version where "startedAt" note was not yet available
        return started == null ? 0 : Integer.parseInt(started);
    }
```

Reply:
We can do it, but I really don't know if it can be problematic. In `AuthenticatedClientSessionEntity` I simply created `getUserSessionStarted` following the same idea that was in `getStarted` (so I returned 0).

For the PR the classes that are really used are `AuthenticatedClientSessionEntity` and `MapAuthenticatedClientSessionEntity` (which don't implement `AuthenticatedClientSessionEntity` and re-implement the access to notes).

Reply:
@mposolda Do you recall why you used 0 as default? To me, this implies that `isIssuedBeforeSessionStart` check will be always `false` meaning "legacy" sessions will always pass this check. At the same time, this PR adds https://github.com/keycloak/keycloak/pull/17525/files#diff-fd8b1c47da5ec7ee42d5450d3f7b73cdeef9c08af26b825c9245e3e6052c5cdfR1090 and https://github.com/keycloak/keycloak/pull/17525/files#diff-fd8b1c47da5ec7ee42d5450d3f7b73cdeef9c08af26b825c9245e3e6052c5cdfR1130 that uses `timestamp` instead 0.

Where I am heading with this, isn't better to use `timestamp` at the first place unless Marek remembers reasons why he used 0?

Reply:
Yes, @martin-kanis, I can give it a try to see it brings any issue. We can change both methods to return the timestamp instead of 0 in `AuthenticatedClientSessionEntity`.

Reply:
@martin-kanis @rmartinc This is good catch!

I recall this was added in this commit https://github.com/keycloak/keycloak/pull/8130/files#diff-b4014a5e9c73d54dc98153e6fa64803e21ab0646a75566cf74ca9a02abbb18db . And some related story behind that is described in this JIRA description https://issues.redhat.com/browse/KEYCLOAK-18368 .

The 0 was just a fallback, which was only used for "offline client sessions" migrated from previous version (As those offline client sessions didn't have "startedAt" note on them). In other words when it would use timestamp instead of 0 in that PR, we wouldn't be able to refresh offline tokens from previous version due this possible scenario:
- Offline token issued on old server (before KEYCLOAK-18368) at time 1000
- Keycloak server stopped and started with new Keycloak version (with KEYCLOAK-18368 applied)
- Refresh of the offline token would fail due this check https://github.com/keycloak/keycloak/blob/21.1.1/services/src/main/java/org/keycloak/protocol/oidc/TokenManager.java#L186-L189

Note that this check is needed exactly because of the scenario described in https://issues.redhat.com/browse/KEYCLOAK-18368. But at the same time, we should make sure that offline tokens from previous version must be possible to refresh. So that's the story of why it was originally added :-)


Now some thoughts for this PR:

IMO the ideal is, if this fallback to 0 (or timestamp) is effectively not needed anywhere.  It looks that the only case how the USER_SESSION_STARTED_AT_NOTE note won't be present is (again) the offline sessions migrated from previous version?

AFAIK we don't need to care about new store as we don't support migrating of offline sessions created with the new store in previous versions. Is it correct @martin-kanis ? So we likely need to care just about old store and the offline sessions coming from `UserSessionPersister`.

So as long as all client sessions returned by `JpaUserSessionPersisterProvider` have this note on them, we should be good. If we always have corresponding UserSession (which AFAIK we have), we can just use `userSession.getStarted()` to fill that note before saving those persisted sessions in infinispan?

Also I wonder if `AuthenticatedClientSessionModel.getUserSessionStarted()` could be updated like this in this PR?

```
    default int getUserSessionStarted() {
        String started = getNote(USER_SESSION_STARTED_AT_NOTE);
        return started == null ? getUserSession().getStarted() : Integer.parseInt(started);
    }
```
In other words, fallback to look at `userSession.getStarted()` instead of using hardcoded value of 0 (or timestamp). Ideal is if AuthenticatedClientSessionEntity also has access to userSession, so it can look into it, but I suppose this would require some bigger refactoring, which we don't want to do in this PR?


In shortcut: In `MigrationTest` we have scenario for testing refresh of the offline-token from previous version. So it seems that this PR should be tested with `MigrationTest` to make sure it works. AFAIK `MigrationTest` is not yet in GH actions, so this may need to be done manually.

Reply:
I have tested this with the `MigrationTest` and it works (`mariadb` only, but I think that's enough). I have also done two little changes:

* The change proposed by @mposolda in `AuthenticatedClientSessionModel.getUserSessionStarted`. 
* And I also modified [these lines](https://github.com/keycloak/keycloak/pull/17525/files#diff-fd8b1c47da5ec7ee42d5450d3f7b73cdeef9c08af26b825c9245e3e6052c5cdfR1090) commented by @martin-kanis. He is right and if 0 the previous check for `isIssuedBeforeSessionStart` fails.

For the previous comments, I think that returning 0 in `AuthenticatedClientSessionModel` interface is necessary. Not changing anything there in this PR for the moment.

WDYT now?

Reply:
@mposolda Thanks for the explanation. Also the proposed changes make sense to me. So @rmartinc LGTM